### PR TITLE
Improved Dijkstra's algorithm.

### PR DESCRIPTION
**Pull request by SyxP**
Modified Dijkstra's algorithm to simplify the code and took care of the case involving negative weight cycle.

### CODE REVIEW

**Comment by SyxP** [comment]
Some benchmarks on large random graphs:
```
julia> G = erdos_renyi(3000,0.78, is_directed = true); AdjMat = rand(3000,3000)

julia> @btime LightGraphs.old_dijkstra($G, 1, $AdjMat)
 216.916 ms (51 allocations: 614.19 KiB)

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat)
 206.399 ms (53 allocations: 614.34 KiB)

julia> G = erdos_renyi(3000,0.05,is_directed=true); AdjMat = rand(3000,3000)

julia> @btime LightGraphs.old_dijkstra($G, 1, $AdjMat)
 18.923 ms (51 allocations: 614.19 KiB)

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat)
 17.746 ms (53 allocations: 614.34 KiB)

julia> G = watts_strogatz(3000,5,0.88,is_directed=true); AdjMat = rand(3000,3000)

julia> @btime LightGraphs.old_dijkstra($G, 1, $AdjMat)
 1.530 ms (40 allocations: 195.80 KiB)

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat)
 1.611 ms (42 allocations: 195.95 KiB)

julia> G = watts_strogatz(3000,1318,0.88,is_directed=true); AdjMat = rand(3000,3000);

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat);
 60.737 ms (53 allocations: 614.34 KiB)

julia> @btime LightGraphs.old_dijkstra($G, 1, $AdjMat);
 71.228 ms (51 allocations: 614.19 KiB)

julia> G = watts_strogatz(100000,10,0.8)
{100000, 500000} undirected simple Int64 graph

julia> @btime dijkstra_shortest_paths($G, 1);
 123.808 ms (72 allocations: 12.58 MiB)

julia> @btime LightGraphs.old_dijkstra($G,1);
 127.472 ms (70 allocations: 12.58 MiB)

```

**Comment by sbromberger** [comment]
This looks promising. What happens when you add the `allpaths=true` and `trackvertices=true` options?

**Comment by SyxP** [comment]
The optional parameters functionality is preserved. Though, `trackvertices` and `allpaths` was never designed to work with graphs with non-positive cycles, and may give unexpected results. 

The code only throws an error on negative weight cycle as only in this case the algorithm will fail. For the case of 0 weight cycles, Dijkstra's will work as per usual.

Finally, regarding the bench marking on large sparse graph.
```
julia> Spars = watts_strogatz(1_000_000,20,0.8)
{1000000, 10000000} undirected simple Int64 graph

julia> using BenchmarkTools

julia> @benchmark LightGraphs.old_dijkstra(Spars,1)
BenchmarkTools.Trial:
  memory estimate:  121.27 MiB
  allocs estimate:  91
  --------------
  minimum time:     2.282 s (3.68% GC)
  median time:      2.520 s (12.17% GC)
  mean time:        2.445 s (9.53% GC)
  maximum time:     2.532 s (12.18% GC)
  --------------
  samples:          3
  evals/sample:     1

julia> @benchmark dijkstra_shortest_paths(Spars,1)
BenchmarkTools.Trial:
  memory estimate:  121.27 MiB
  allocs estimate:  93
  --------------
  minimum time:     2.137 s (3.88% GC)
  median time:      2.333 s (13.16% GC)
  mean time:        2.274 s (10.26% GC)
  maximum time:     2.351 s (13.06% GC)
  --------------
  samples:          3
  evals/sample:     1
```

File: src/shortestpaths/dijkstra.jl [hunk 1]
```diff
@@ -66,51 +70,47 @@ function dijkstra_shortest_paths(g::AbstractGraph,
...
 
     closest_vertices = Vector{U}()  # Maintains vertices in order of distances from source
     sizehint!(closest_vertices, nvg)
 
-    while !isempty(H)
-        u = dequeue!(H)
+    while !isempty(pq)
+        u, d = peek(pq) #(u, distance of u)
+        dists[u] = d
+        dequeue!(pq)
```
[the remark this answers is not in the document]
Comment (SyxP) [reply src/shortestpaths/dijkstra.jl]: I didn't know of such a function. This has been added.

**Comment by sbromberger** [comment]
What do the new benchmarks look like?

**Comment by SyxP** [comment]
In comparison to the above:
```
julia> G = erdos_renyi(3000,0.78, is_directed = true); AdjMat = rand(3000,3000);

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat)
  141.036 ms (53 allocations: 614.34 KiB)

julia> G = watts_strogatz(3000,1318,0.88,is_directed=true); AdjMat = rand(3000,3000);

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat)
  42.590 ms (53 allocations: 614.34 KiB)

julia> Spars = watts_strogatz(1_000_000,20,0.8)
{1000000, 10000000} undirected simple Int64 graph

julia>  @benchmark dijkstra_shortest_paths(Spars,1)
BenchmarkTools.Trial:
  memory estimate:  121.27 MiB
  allocs estimate:  93
  --------------
  minimum time:     1.594 s (4.38% GC)
  median time:      1.764 s (13.74% GC)
  mean time:        1.708 s (10.87% GC)
  maximum time:     1.765 s (13.85% GC)
  --------------
  samples:          3
  evals/sample:     1
```

**Comment by sbromberger** [comment]
that's significant. Nice.

**Comment by SyxP** [comment]
Two key differences have been made. 

1) Previously, I attempted to handle negative weights (wrongly). I could fix it but I realise that this might lead to unexpected exponential time algorithm (as Dijkstra was never meant to work with negative weights). 

2) A priority queue implementation has been changed. Let me discuss the tradeoff here. Speed-wise, there should be no significant loss in time. However, on dense graphs, as each vertex may be pushed in more than once, this may lead to increase in memory usage. The additional memory usage is capped at $O(E)$ now rather than previously at $O(V)$. (It is possible to design a testcase with this explicity memory goal in mind. But I think this does not occur sufficiently common in practice and justifies this change.) Perhaps, we should modify the code to take a Priority Queue as an optional parameter, but there is no unified interface for that. (That may also solve the problem of undirected graph, as we can pass a Queue as a Priority Queue when no distmx have been passed in.) 

Now for some benchmarks.

```
julia> G = erdos_renyi(3000,0.78, is_directed = true); AdjMat = rand(3000,3000);

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat);
  159.984 ms (31 allocations: 1.12 MiB)

julia> G = watts_strogatz(3000,1318,0.88,is_directed=true); AdjMat = rand(3000,3000);

julia> @btime dijkstra_shortest_paths($G, 1, $AdjMat);
  52.852 ms (30 allocations: 633.80 KiB)

julia> Spars = watts_strogatz(1_000_000,20,0.8);

julia> b = @benchmarkable dijkstra_shortest_paths(Spars, 1) seconds = 100
Benchmark(evals=1, seconds=100.0, samples=10000)

julia> run(b)
BenchmarkTools.Trial:
  memory estimate:  56.10 MiB
  allocs estimate:  39
  --------------
  minimum time:     1.169 s (0.60% GC)
  median time:      1.199 s (0.58% GC)
  mean time:        1.252 s (0.61% GC)
  maximum time:     1.698 s (0.39% GC)
  --------------
  samples:          80
  evals/sample:     1
```

File: test/shortestpaths/dijkstra.jl [hunk 2]
```diff
@@ -81,7 +81,7 @@
         dm = @inferred(dijkstra_shortest_paths(g, 1; allpaths=true, trackvertices=true))
         @test dm.pathcounts       == [1, 1, 1, 1, 2]
         @test dm.predecessors     == [[], [1], [1], [3], [2, 3]]
-        @test dm.closest_vertices == [1, 2, 3, 5, 4]
+        @test dm.closest_vertices == [1, 2, 3, 4, 5]
     end
```
Comment (sbromberger) [inline test/shortestpaths/dijkstra.jl]: be very careful here. `closest_vertices` has a specific order in order for `betweenness_centrality` to work properly.

Comment (SyxP) [reply test/shortestpaths/dijkstra.jl]: The issue is that they are both the same distance apart, and it is dependent on the type of priority queue to determine which one gets inserted first.

Comment (sbromberger) [reply test/shortestpaths/dijkstra.jl]: Then you can't change it. Sorry. I think too many things are dependent on a specific order. Check with @somil55 on the details.

Comment (SyxP) [reply test/shortestpaths/dijkstra.jl]: For the specific case of `between_centrality`, if the comment there is anything to go by
```
# we need to order the source vertices by decreasing distance for this to work.
S = reverse(state.closest_vertices) #Replaced sortperm with this
```
suggest that for vertices on same distance, their order shouldn't matter. Another way to think about it is that even previously (using the PriorityQueue) on a same graph, by permuting the vertex labels, I can achieve different outcomes of this array.

**Comment by SohamTamba** [comment]
Minimising memory usage usually leads to better performance because it tends to use cache more efficiently.
Also I'm fairly sure efficient Dijkstra implementations usually use PriorityQueue.

Is it possible that this discrepancy is due to the implementations of PriorityQueue and MinHeap?
For example, the current PriorityQueue uses a Hash Table to keep track of the indices. Since LG indexes vertices in `1:nv(g)` it could be replaced with an array.

**Comment by SyxP** [comment]
The main issue does lie with `DataStructures.jl` as they use an additional dictionary to map my vertex labels to integers. This was why I opted for using a `MinHeap` to optimise it further. Moreover, there have been various studies done (I have yet to replicate any of it, but I read a couple of sources like https://pdfs.semanticscholar.org/2453/8637bd3c9f6973003370dd176f4b0b80f8f1.pdf) that claims that the decrease-key operation is very costly, and may only be beneficial in the worst case. 

Next, regarding the vertex labels, this is indeed true, and I'm unsure of why `vertices(g)` doesn't just return `1:nv(g)`.

**Comment by sbromberger** [comment]
> I'm unsure of why vertices(g) doesn't just return `1:nv(g)`.

It does for SimpleGraphs: 
https://github.com/JuliaGraphs/LightGraphs.jl/blob/3f71acda3996c429f5df268649f05b79e85def62/src/SimpleGraphs/SimpleGraphs.jl#L56

**Comment by SohamTamba** [comment]
I benchmarked Dijkstra with the 3 datastructues and surprisingly, heap and integer priority queue have almost the same performance.
See: [Branch Int_PQ](https://github.com/SohamTamba/LightGraphs.jl/blob/master/src/shortestpaths/dijkstra.jl)
```
|V| = 81306        |E| = 1342310

--------------------PQ DIJKSTRA--------------------------
  501.488 ms (67 allocations: 10.85 MiB)
--------------------int_PQ DIJKSTRA------------------------------
  414.363 ms (33 allocations: 5.80 MiB)
--------------------HEAP DIJKSTRA------------------------------
  407.132 ms (35 allocations: 8.18 MiB)
```
It would be better to stick to @SyxP 's implementation.
Btw the negative cycle error increases the runtime alot (400ms to 600ms).

Dijkstra is supposed have only non-negative weights as the input, so there should be an easier way to resolve this.
For example, 
1. Throw an error if any edge has negative weight.
2. Replace `distmx[u, v]` with `max(distmx[u, v], 0)`

**Comment by simonschoelly** [comment]
What graphs did you use for benchmarking?

And aren't priority ques usually implemented using a heap anyways? Then I would not be surprised that they have the same running time.

**Comment by SyxP** [comment]
The difference here is that the `IntPriorityQueue` ensures that the same vertex wouldn't be in the priority queue more than once, while a Heap implementation would not have such a gurantee. The tradeoff here is that the Heap implementation does not need to deal with decreasekey, while the `IntPriorityQueue` can ensure the memory is small. 

On checking whether an edge is negative, it is possible for there to be a negative edge, but not affect the result. For instance, it may be disconnected from the part of the graph we're searching from. I have no good answers regarding how this should be done and open to editing it.

**Comment by SohamTamba** [comment]
> What graphs did you use for benchmarking?
> 
> And aren't priority ques usually implemented using a heap anyways? Then I would not be surprised that they have the same running time.

I'm using the Twitter Graph from SNAPDatasets.

**Comment by SohamTamba** [comment]
> The difference here is that the `IntPriorityQueue` ensures that the same vertex wouldn't be in the priority queue more than once, while a Heap implementation would not have such a gurantee. The tradeoff here is that the Heap implementation does not need to deal with decreasekey, while the `IntPriorityQueue` can ensure the memory is small.
> 
> On checking whether an edge is negative, it is possible for there to be a negative edge, but not affect the result. For instance, it may be disconnected from the part of the graph we're searching from. I have no good answers regarding how this should be done and open to editing it.

I think it should left to the user to make sure even disconnected components do not have negative edges.

Should I open an issue for this? Many algos like dijkstra, spanning tree, steiner tree etc. make assumptions about connectivity and edge weights.

Either I think we can merge this code (without the negative weight error IMO) and decide later what to do for these assumptions.

**Comment by simonschoelly** [comment]
As this algorithm is the workhorse of a lot of other algorithms and has to perform fast, I think it is reasonable to assume that it is upper to the user, not to use negative weights. This assumption has to be clearly stated in the docstring though.

**Comment by sbromberger** [comment]
Let's get this fixed up and merged.

**Pull request closed without merging**